// ts-patch/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLiveModule = exports.check = exports.patch = exports.uninstall = exports.install = exports.getInstallerOptions = exports.InstallerOptions = void 0;
var options_1 = require("./options");
Object.defineProperty(exports, "InstallerOptions", { enumerable: true, get: function () { return options_1.InstallerOptions; } });
Object.defineProperty(exports, "getInstallerOptions", { enumerable: true, get: function () { return options_1.getInstallerOptions; } });
var actions_1 = require("./actions");
Object.defineProperty(exports, "install", { enumerable: true, get: function () { return actions_1.install; } });
Object.defineProperty(exports, "uninstall", { enumerable: true, get: function () { return actions_1.uninstall; } });
Object.defineProperty(exports, "patch", { enumerable: true, get: function () { return actions_1.patch; } });
Object.defineProperty(exports, "check", { enumerable: true, get: function () { return actions_1.check; } });
var module_1 = require("./module");
Object.defineProperty(exports, "getLiveModule", { enumerable: true, get: function () { return module_1.getLiveModule; } });
__exportStar(require("./plugin-types"), exports);
